### PR TITLE
catalog search service honor experiment flag

### LINKED ISSUE #3355: odo search service doesn't honor experiment flag
/kind bug

## What versions of software are you using?

**Operating System:**

**Output of `odo version`:**
odo v1.2.2 (a15a99e)

Server: https://api.crc.testing:6443
Kubernetes: v1.17.1

## How did you run odo exactly?
```
 odo catalog search service mysql
 ✗  unable to list services because Service Catalog is not enabled in your cluster: unable to list services: unable to get cluster serviceClassExternalName: unable to get cluster service classes: unable to list cluster service classes: the server could not find the requested resource (get clusterserviceclasses.servicecatalog.k8s.io)
[jichen@jjtest2 nodejs-ex]$ odo preference set Experimental true
? Experimental is already set. Do you want to override it in the config Yes
Global preference was successfully updated
[jichen@jjtest2 nodejs-ex]$ odo catalog search service mysql
 ✗  unable to list services because Service Catalog is not enabled in your cluster: unable to list services: unable to get cluster serviceClassExternalName: unable to get cluster service classes: unable to list cluster service classes: the server could not find the requested resource (get clusterserviceclasses.servicecatalog.k8s.io)
```

the list command provided different output when Experiment flag is set or not

### PR DESCRIPTION
**Pull request by jichenjc**
**What type of PR is this?**

> /kind feature
**What does does this PR do / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #3355

**How to test changes / Special notes to the reviewer**:
`odo catalog search service <>`

### CODE REVIEW

**Comment by jichenjc** [comment]
ci/prow/v4.4-integration-e2e

**Comment by jichenjc** [comment]
@dharmit can you help to provide some comments? Thanks

**Comment by dharmit** [comment]
@jichenjc Thanks for this PR!

Here's the list of Operators and available CRDs in my cluster:
```sh
$ odo catalog list services
Operators available in the cluster
NAME                                    CRDs
cockroachdb.v3.0.7                      Cockroachdb
etcdoperator.v0.9.4-clusterwide         EtcdCluster, EtcdBackup, EtcdRestore
postgresql-operator.v0.1.1              Backup, Database
service-binding-operator.v0.1.1-282     ServiceBindingRequest
```
And here's the output of some search queries:
```sh
$ odo catalog search service postgres
Services available through Service Catalog
NAME     PLANS
Operators available in the cluster
NAME                           CRDs
postgresql-operator.v0.1.1     Backup, Database

$ odo catalog search service etcd    
Services available through Service Catalog
NAME     PLANS
Operators available in the cluster
NAME                                CRDs
etcdoperator.v0.9.4-clusterwide     EtcdCluster, EtcdBackup, EtcdRestore

$ odo catalog search service Etcd
 ✗  no service matched the query: Etcd

$ odo catalog search service EtcdCluster
 ✗  no service matched the query: EtcdCluster
```

Few requests/suggestions:
1. Can we ensure that we don't print below blurb:
    ```
    Services available through Service Catalog
    NAME     PLANS
    ```
2. Can we please search not just based on Operator name but also CRD names. For example, `odo catalog search service Etcd` & `odo catalog search service EtcdCluster` should have shown:
    ```
    $ odo catalog search service etcd    
    Operators available in the cluster
    NAME                                CRDs
    etcdoperator.v0.9.4-clusterwide     EtcdCluster, EtcdBackup, EtcdRestore
    ```

**Comment by jichenjc** [comment]
@dharmit thanks for the reminder, I updated the PR and tested locally

File: pkg/kclient/operators.go [hunk 1]
```diff
@@ -41,3 +48,29 @@ func (c *Client) GetCustomResourcesFromCSV(csv olm.ClusterServiceVersion) []olm.
 	// we will return a list of CRs owned by the csv
 	return csv.Spec.CustomResourceDefinitions.Owned
 }
+
+// SearchClusterServiceVersionList searches for the CSV
```
Comment (dharmit) [inline pkg/kclient/operators.go:52]: Please add more comment here that indicates that this works with the Operators.

File: pkg/odo/cli/catalog/search/service.go [hunk 2]
```diff
@@ -29,30 +32,74 @@ func NewSearchServiceOptions() *SearchServiceOptions {
 
 // Complete completes SearchServiceOptions after they've been created
 func (o *SearchServiceOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
-	o.Context = genericclioptions.NewContext(cmd)
-	o.searchTerm = args[0]
+	if experimental.IsExperimentalModeEnabled() {
+		var noCsvs, noServices bool
+		o.Context = genericclioptions.NewContext(cmd)
```
Comment (dharmit) [inline pkg/odo/cli/catalog/search/service.go:37]: In case of experimental mode being enabled, we use `NewDevfileContext`.

**Comment by dharmit** [comment]
@jichenjc Thanks for making the changes!

Besides the review, can you please also make sure that search is case-insensitive so that both of the below queries would work:
```sh
$ odo catalog search service rest           
 ✗  no service matched the query: rest

$ odo catalog search service Rest
Operators available in the cluster
NAME                                CRDs
etcdoperator.v0.9.4-clusterwide     EtcdCluster, EtcdBackup, EtcdRestore
```

**Comment by jichenjc** [comment]
> 
> 
> @jichenjc Thanks for making the changes!
> 
> Besides the review, can you please also make sure that search is case-insensitive so that both of the below queries would work:
> 
> ```shell
> $ odo catalog search service rest           
>  ✗  no service matched the query: rest
> 
> $ odo catalog search service Rest
> Operators available in the cluster
> NAME                                CRDs
> etcdoperator.v0.9.4-clusterwide     EtcdCluster, EtcdBackup, EtcdRestore
> ```

@dharmit can you help to confirm with service catalog (experiment = false)
you see the previous search result is case-insensitive? 
from below code 
https://github.com/openshift/odo/blob/master/pkg/catalog/catalog.go#L257
I think it's case sensitive.. I don't have service catalog env for me to verify. thanks

File: pkg/odo/cli/catalog/search/service.go (hunk 2 above)
Comment (jichenjc) [reply pkg/odo/cli/catalog/search/service.go]: done

File: pkg/kclient/operators.go (hunk 1 above)
Comment (jichenjc) [reply pkg/kclient/operators.go]: done

File: pkg/odo/cli/catalog/search/service.go (hunk 2 above)
Comment (jichenjc) [reply pkg/odo/cli/catalog/search/service.go]: @dharmit can't do so...
if you check below, actually 

it calls non-experiment code
```
o.services, err = catalog.SearchService(o.Client, o.searchTerm)
```
so this has to be NewContext

**Comment by jichenjc** [comment]
https://ibm.ent.box.com/notes/680378804271

**Comment by jichenjc** [comment]
@dharmit please let me know any comments regarding above ... thanks

**Comment by dharmit** [comment]
> I think it's case sensitive and I think we need be consistent
> 
> I don't have service catalog env for me to verify.... thanks

You're right. It's case-sensitive. Let's keep the behaviour consistent. :+1:

File: pkg/odo/cli/catalog/util/util.go [hunk 3]
```diff
@@ -69,12 +69,12 @@ func DisplayClusterServiceVersions(csvs *olm.ClusterServiceVersionList) {
...
 	fmt.Fprintln(w, "NAME", "\t", "CRDs")
 	for _, csv := range csvs.Items {
-		fmt.Fprintln(w, csv.ObjectMeta.Name, "\t", csvOperators(csv.Spec.CustomResourceDefinitions))
+		fmt.Fprintln(w, csv.ObjectMeta.Name, "\t", CsvOperators(csv.Spec.CustomResourceDefinitions))
 	}
 	w.Flush()
 }
 
-func csvOperators(crds olm.CustomResourceDefinitions) string {
+func CsvOperators(crds olm.CustomResourceDefinitions) string {
```
Comment (dharmit) [inline pkg/odo/cli/catalog/util/util.go:77]: Why are we making this function exported? Please add a comment if it needs to stay exported.

Comment (jichenjc) [reply pkg/odo/cli/catalog/util/util.go:77]: yes, I will add a comment, it's need to be exported

**Comment by dharmit** [comment]
@mik-dass @kadel @girishramnani @cdrage can one of you approve this pr? I've tried the functionality and it looks good to me.

**Pull request merged**